File: rust/wxdragon-sys/cpp/src/bitmapbundle.cpp
#include "../include/wxdragon.h"
#include <wx/wx.h>
#include <wx/bmpbndl.h>
#include <wx/bitmap.h>
#include <wx/mstream.h>
#include <vector>

// Create an empty bitmap bundle
WXD_EXPORTED wxd_BitmapBundle_t* wxd_BitmapBundle_Create() {
    wxBitmapBundle* bundle = new wxBitmapBundle();
    return reinterpret_cast<wxd_BitmapBundle_t*>(bundle);
}

// Create a bitmap bundle from a single bitmap
WXD_EXPORTED wxd_BitmapBundle_t* wxd_BitmapBundle_CreateFromBitmap(wxd_Bitmap_t* bitmap) {
    if (!bitmap) {
        return wxd_BitmapBundle_Create();
    }
    
    wxBitmap* wx_bitmap = reinterpret_cast<wxBitmap*>(bitmap);
    wxBitmapBundle* bundle = new wxBitmapBundle(*wx_bitmap);
    return reinterpret_cast<wxd_BitmapBundle_t*>(bundle);
}

// Create a bitmap bundle from multiple bitmaps of different sizes
WXD_EXPORTED wxd_BitmapBundle_t* wxd_BitmapBundle_FromBitmaps(wxd_Bitmap_t** bitmaps, size_t count) {
    if (!bitmaps || count == 0) {
        return wxd_BitmapBundle_Create();
    }
    
    wxVector<wxBitmap> wx_bitmaps;
    wx_bitmaps.reserve(count);
    
    for (size_t i = 0; i < count; ++i) {
        if (bitmaps[i]) {
            wxBitmap* wx_bitmap = reinterpret_cast<wxBitmap*>(bitmaps[i]);
            wx_bitmaps.push_back(*wx_bitmap);
        }
    }
    
    if (wx_bitmaps.empty()) {
        return wxd_BitmapBundle_Create();
    }
    
    wxBitmapBundle* bundle = new wxBitmapBundle(wxBitmapBundle::FromBitmaps(wx_bitmaps));
    return reinterpret_cast<wxd_BitmapBundle_t*>(bundle);
}

// Clone a bitmap bundle
WXD_EXPORTED wxd_BitmapBundle_t* wxd_BitmapBundle_Clone(const wxd_BitmapBundle_t* bundle) {
    if (!bundle) {
        return wxd_BitmapBundle_Create();
    }
    
    const wxBitmapBundle* wx_bundle = reinterpret_cast<const wxBitmapBundle*>(bundle);
    wxBitmapBundle* new_bundle = new wxBitmapBundle(*wx_bundle);
    return reinterpret_cast<wxd_BitmapBundle_t*>(new_bundle);
}

// Destroy a bitmap bundle
WXD_EXPORTED void wxd_BitmapBundle_Destroy(wxd_BitmapBundle_t* bundle) {
    if (bundle) {
        wxBitmapBundle* wx_bundle = reinterpret_cast<wxBitmapBundle*>(bundle);
        delete wx_bundle;
    }
}

// Create a bitmap bundle from an SVG file
WXD_EXPORTED wxd_BitmapBundle_t* wxd_BitmapBundle_FromSVGFile(const char* path, wxd_Size size) {
    if (!path) {
        return wxd_BitmapBundle_Create();
    }
    
    wxString wx_path = wxString::FromUTF8(path);
    wxSize wx_size(size.width, size.height);
    wxBitmapBundle* bundle = new wxBitmapBundle(wxBitmapBundle::FromSVGFile(wx_path, wx_size));
    return reinterpret_cast<wxd_BitmapBundle_t*>(bundle);
}

// Create a bitmap bundle from SVG text
WXD_EXPORTED wxd_BitmapBundle_t* wxd_BitmapBundle_FromSVGText(const char* svg_text, wxd_Size size) {
    if (!svg_text) {
        return wxd_BitmapBundle_Create();
    }
    
    wxString wx_svg = wxString::FromUTF8(svg_text);
    wxSize wx_size(size.width, size.height);
    wxBitmapBundle* bundle = new wxBitmapBundle(wxBitmapBundle::FromSVG(wx_svg, wx_size));
    return reinterpret_cast<wxd_BitmapBundle_t*>(bundle);
}

// Create a bitmap bundle from SVG data
WXD_EXPORTED wxd_BitmapBundle_t* wxd_BitmapBundle_FromSVGData(const unsigned char* data, size_t len, wxd_Size size) {
    if (!data || len == 0) {
        return wxd_BitmapBundle_Create();
    }
    
    wxSize wx_size(size.width, size.height);
    wxBitmapBundle* bundle = new wxBitmapBundle(wxBitmapBundle::FromSVG(data, len, wx_size));
    return reinterpret_cast<wxd_BitmapBundle_t*>(bundle);
}

// Get a bitmap of the specified size from the bundle
WXD_EXPORTED wxd_Bitmap_t* wxd_BitmapBundle_GetBitmap(const wxd_BitmapBundle_t* bundle, wxd_Size size) {
    if (!bundle) {
        return nullptr;
    }
    
    const wxBitmapBundle* wx_bundle = reinterpret_cast<const wxBitmapBundle*>(bundle);
    wxSize wx_size(size.width, size.height);
    wxBitmap bitmap = wx_bundle->GetBitmap(wx_size);
    
    if (!bitmap.IsOk()) {
        return nullptr;
    }
    
    wxBitmap* new_bitmap = new wxBitmap(bitmap);
    return reinterpret_cast<wxd_Bitmap_t*>(new_bitmap);
}

// Get a bitmap for a specific window from the bundle
WXD_EXPORTED wxd_Bitmap_t* wxd_BitmapBundle_GetBitmapFor(const wxd_BitmapBundle_t* bundle, wxd_Window_t* window) {
    if (!bundle || !window) {
        return nullptr;
    }
    
    const wxBitmapBundle* wx_bundle = reinterpret_cast<const wxBitmapBundle*>(bundle);
    wxWindow* wx_window = reinterpret_cast<wxWindow*>(window);
    wxBitmap bitmap = wx_bundle->GetBitmapFor(wx_window);
    
    if (!bitmap.IsOk()) {
        return nullptr;
    }
    
    wxBitmap* new_bitmap = new wxBitmap(bitmap);
    return reinterpret_cast<wxd_Bitmap_t*>(new_bitmap);
}

// Get the default size of the bitmap bundle
WXD_EXPORTED wxd_Size wxd_BitmapBundle_GetDefaultSize(const wxd_BitmapBundle_t* bundle) {
    wxd_Size size = {0, 0};
    
    if (bundle) {
        const wxBitmapBundle* wx_bundle = reinterpret_cast<const wxBitmapBundle*>(bundle);
        wxSize wx_size = wx_bundle->GetDefaultSize();
        size.width = wx_size.GetWidth();
        size.height = wx_size.GetHeight();
    }
    
    return size;
}

// Get the preferred bitmap size at a specific scale
WXD_EXPORTED wxd_Size wxd_BitmapBundle_GetPreferredBitmapSizeAtScale(const wxd_BitmapBundle_t* bundle, double scale) {
    wxd_Size size = {0, 0};
    
    if (bundle) {
        const wxBitmapBundle* wx_bundle = reinterpret_cast<const wxBitmapBundle*>(bundle);
        wxSize wx_size = wx_bundle->GetPreferredBitmapSizeAtScale(scale);
        size.width = wx_size.GetWidth();
        size.height = wx_size.GetHeight();
    }
    
    return size;
}

// Get the preferred bitmap size for a specific window
WXD_EXPORTED wxd_Size wxd_BitmapBundle_GetPreferredBitmapSizeFor(const wxd_BitmapBundle_t* bundle, wxd_Window_t* window) {
    wxd_Size size = {0, 0};
    
    if (bundle && window) {
        const wxBitmapBundle* wx_bundle = reinterpret_cast<const wxBitmapBundle*>(bundle);
        wxWindow* wx_window = reinterpret_cast<wxWindow*>(window);
        wxSize wx_size = wx_bundle->GetPreferredBitmapSizeFor(wx_window);
        size.width = wx_size.GetWidth();
        size.height = wx_size.GetHeight();
    }
    
    return size;
}

// Check if the bitmap bundle is valid
WXD_EXPORTED bool wxd_BitmapBundle_IsOk(const wxd_BitmapBundle_t* bundle) {
    if (!bundle) {
        return false;
    }
    
    const wxBitmapBundle* wx_bundle = reinterpret_cast<const wxBitmapBundle*>(bundle);
    return wx_bundle->IsOk();
} 

File: rust/wxdragon/src/bitmap_bundle.rs
//!
//! Safe wrapper for wxBitmapBundle.

use std::path::Path;
use std::ffi::CString;
use wxdragon_sys as ffi;

use crate::bitmap::Bitmap;
use crate::geometry::Size;
use crate::window::WxWidget;

/// Represents a collection of bitmaps of the same image in different sizes/resolutions.
///
/// This class allows the application to provide images in different resolutions so
/// that they can be displayed optimally at different DPI levels. It's especially useful
/// for toolbar icons and other UI elements that need to scale well at high DPI.
///
/// BitmapBundle can be created from:
/// - A single bitmap (for backwards compatibility)
/// - Multiple bitmaps of different sizes (for optimal display at different DPI levels)
/// - An SVG file (which can be rendered at any size)
/// - SVG data in memory
/// - SVG string content
#[derive(Debug)]
pub struct BitmapBundle {
    ptr: *mut ffi::wxd_BitmapBundle_t,
    is_owned: bool, // Tracks whether Rust owns this bundle and should destroy it
}

impl BitmapBundle {
    /// Creates a new empty bitmap bundle.
    ///
    /// While valid, the empty bundle is not particularly useful as it doesn't
    /// contain any bitmaps to be shown.
    pub fn new() -> Self {
        let ptr = unsafe { ffi::wxd_BitmapBundle_Create() };
        BitmapBundle {
            ptr,
            is_owned: true,
        }
    }

    /// Creates a bitmap bundle from a single bitmap.
    ///
    /// This is primarily for backward compatibility with APIs that used to take a wxBitmap.
    pub fn from_bitmap(bitmap: &Bitmap) -> Self {
        let ptr = unsafe { ffi::wxd_BitmapBundle_CreateFromBitmap(bitmap.as_ptr()) };
        BitmapBundle {
            ptr,
            is_owned: true,
        }
    }

    /// Creates a bitmap bundle from multiple bitmaps of different sizes.
    ///
    /// This is the preferred way to create a bitmap bundle for applications supporting
    /// multiple DPIs. Provide the same image in different resolutions to allow wxWidgets
    /// to choose the most appropriate one for the current display.
    ///
    /// # Arguments
    /// * `bitmaps` - A vector of bitmaps to include in the bundle. For best results,
    ///   include versions at 100%, 150%, and 200% of the base size (e.g., 16x16, 24x24, 32x32).
    ///
    /// # Returns
    /// A new bitmap bundle containing all the provided bitmaps.
    /// If the vector is empty, returns an empty bundle.
    pub fn from_bitmaps(bitmaps: &[Bitmap]) -> Self {
        if bitmaps.is_empty() {
            return Self::new();
        }
        
        // Create an array of pointers to the bitmaps
        let mut bitmap_ptrs: Vec<*mut ffi::wxd_Bitmap_t> = bitmaps
            .iter()
            .map(|bmp| bmp.as_ptr())
            .collect();
        
        let ptr = unsafe { 
            ffi::wxd_BitmapBundle_FromBitmaps(bitmap_ptrs.as_mut_ptr(), bitmaps.len())
        };
        
        BitmapBundle {
            ptr,
            is_owned: true,
        }
    }

    /// Creates a bitmap bundle from an SVG file.
    ///
    /// # Arguments
    /// * `path` - Path to the SVG file.
    /// * `default_size` - The size to use when the exact size needed is not known.
    ///
    /// # Returns
    /// None if the file could not be loaded or if the SVG is invalid.
    pub fn from_svg_file<P: AsRef<Path>>(path: P, default_size: Size) -> Option<Self> {
        let c_path = match path.as_ref().to_str().map(CString::new) {
            Some(Ok(s)) => s,
            _ => return None,
        };
        
        let size: ffi::wxd_Size = default_size.into();
        
        let ptr = unsafe { ffi::wxd_BitmapBundle_FromSVGFile(c_path.as_ptr(), size) };
        if ptr.is_null() {
            None
        } else {
            Some(BitmapBundle {
                ptr,
                is_owned: true,
            })
        }
    }

    /// Creates a bitmap bundle from SVG text.
    ///
    /// # Arguments
    /// * `svg_text` - String containing SVG content.
    /// * `default_size` - The size to use when the exact size needed is not known.
    ///
    /// # Returns
    /// None if the SVG content is invalid.
    pub fn from_svg_text(svg_text: &str, default_size: Size) -> Option<Self> {
        let c_svg = match CString::new(svg_text) {
            Ok(s) => s,
            Err(_) => return None,
        };
        
        let size: ffi::wxd_Size = default_size.into();
        
        let ptr = unsafe { ffi::wxd_BitmapBundle_FromSVGText(c_svg.as_ptr(), size) };
        if ptr.is_null() {
            None
        } else {
            Some(BitmapBundle {
                ptr,
                is_owned: true,
            })
        }
    }

    /// Creates a bitmap bundle from raw SVG data.
    ///
    /// # Arguments
    /// * `data` - Byte slice containing SVG data.
    /// * `default_size` - The size to use when the exact size needed is not known.
    ///
    /// # Returns
    /// None if the SVG data is invalid.
    pub fn from_svg_data(data: &[u8], default_size: Size) -> Option<Self> {
        if data.is_empty() {
            return None;
        }
        
        let size: ffi::wxd_Size = default_size.into();
        
        let ptr = unsafe { 
            ffi::wxd_BitmapBundle_FromSVGData(
                data.as_ptr(),
                data.len(),
                size
            ) 
        };
        
        if ptr.is_null() {
            None
        } else {
            Some(BitmapBundle {
                ptr,
                is_owned: true,
            })
        }
    }

    /// Retrieves a bitmap of the specified size from the bundle.
    ///
    /// The returned bitmap is the best match for the requested size, possibly scaled
    /// from one of the bitmaps in the bundle or rendered from SVG if the bundle was
    /// created from SVG content.
    ///
    /// # Arguments
    /// * `size` - Desired size in pixels.
    ///
    /// # Returns
    /// None if the bundle is invalid or couldn't create a bitmap.
    pub fn get_bitmap(&self, size: Size) -> Option<Bitmap> {
        let wxd_size: ffi::wxd_Size = size.into();
        
        let bitmap_ptr = unsafe { ffi::wxd_BitmapBundle_GetBitmap(self.ptr, wxd_size) };
        
        if bitmap_ptr.is_null() {
            None
        } else {
            // The C++ side returns a new bitmap that we own
            Some(Bitmap::from_ptr_owned(bitmap_ptr))
        }
    }

    /// Retrieves a bitmap sized appropriately for the DPI scaling of the given window.
    ///
    /// This is the recommended way to get a bitmap from the bundle when you know
    /// which window it will be displayed in, as it ensures the bitmap is sized
    /// correctly for the window's DPI.
    ///
    /// # Arguments
    /// * `window` - The window where the bitmap will be displayed.
    ///
    /// # Returns
    /// None if the bundle is invalid, the window is invalid, or a bitmap couldn't be created.
    pub fn get_bitmap_for(&self, window: &dyn WxWidget) -> Option<Bitmap> {
        let bitmap_ptr = unsafe { 
            ffi::wxd_BitmapBundle_GetBitmapFor(self.ptr, window.handle_ptr())
        };
        
        if bitmap_ptr.is_null() {
            None
        } else {
            // The C++ side returns a new bitmap that we own
            Some(Bitmap::from_ptr_owned(bitmap_ptr))
        }
    }

    /// Gets the default size of the bundle.
    ///
    /// This is the size that the bundle considers its "standard" size,
    /// typically the size of the original bitmap or the default size specified
    /// when creating from SVG.
    pub fn get_default_size(&self) -> Size {
        let size = unsafe { ffi::wxd_BitmapBundle_GetDefaultSize(self.ptr) };
        Size::from(size)
    }

    /// Gets the preferred bitmap size for a specific scale factor.
    ///
    /// This returns the size of the bitmap that would be best to use at
    /// the given scale factor (e.g., 1.0 for standard DPI, 2.0 for 200% scaling).
    ///
    /// # Arguments
    /// * `scale` - The scale factor (1.0 = 100%, 2.0 = 200%, etc.)
    pub fn get_preferred_bitmap_size_at_scale(&self, scale: f64) -> Size {
        let size = unsafe { 
            ffi::wxd_BitmapBundle_GetPreferredBitmapSizeAtScale(self.ptr, scale)
        };
        Size::from(size)
    }

    /// Gets the preferred bitmap size for a specific window.
    ///
    /// This returns the size of the bitmap that would be best to use for
    /// the DPI scaling of the given window.
    ///
    /// # Arguments
    /// * `window` - The window where the bitmap will be displayed.
    pub fn get_preferred_bitmap_size_for(&self, window: &dyn WxWidget) -> Size {
        let size = unsafe { 
            ffi::wxd_BitmapBundle_GetPreferredBitmapSizeFor(self.ptr, window.handle_ptr())
        };
        Size::from(size)
    }

    /// Checks if the bitmap bundle is valid and contains at least one bitmap.
    pub fn is_ok(&self) -> bool {
        !self.ptr.is_null() && unsafe { ffi::wxd_BitmapBundle_IsOk(self.ptr) }
    }

    /// Returns the raw bundle pointer.
    /// Use with caution, primarily for internal FFI calls.
    pub fn as_ptr(&self) -> *mut ffi::wxd_BitmapBundle_t {
        self.ptr
    }
}

impl Clone for BitmapBundle {
    fn clone(&self) -> Self {
        unsafe {
            let cloned_ptr = ffi::wxd_BitmapBundle_Clone(self.ptr);
            if cloned_ptr.is_null() {
                panic!(
                    "Failed to clone wxBitmapBundle: wxd_BitmapBundle_Clone returned null. Original: {:?}",
                    self.ptr
                );
            }
            // A cloned bundle is always owned by Rust
            BitmapBundle {
                ptr: cloned_ptr,
                is_owned: true,
            }
        }
    }
}

impl Drop for BitmapBundle {
    fn drop(&mut self) {
        if !self.ptr.is_null() && self.is_owned {
            unsafe {
                ffi::wxd_BitmapBundle_Destroy(self.ptr);
            }
        }
    }
}

impl Default for BitmapBundle {
    fn default() -> Self {
        Self::new()
    }
} 